src/hosts.cpp

```diff
@@ -85,6 +88,43 @@ code hosts::fetch(address& out) const
     ///////////////////////////////////////////////////////////////////////////
 }
 
+code hosts::fetch(address::list& out) const
+{
+    if (disabled_)
+        return error::not_found;
+
+    size_t out_count = 0;
+
+    // Critical Section
+    ///////////////////////////////////////////////////////////////////////////
+    {
+        shared_lock lock(mutex_);
+
+        if (stopped_)
+            return error::service_stopped;
+
+        if (buffer_.empty())
+            return error::not_found;
+
+        out_count = std::min(buffer_.size(), capacity_) /
+            static_cast<size_t>(pseudo_random(1, 20));
+    }
+    ///////////////////////////////////////////////////////////////////////////
+
+    if (out_count == 0)
+        return error::success;
+
+    out.reserve(out_count);
+
+    for (size_t index = 0; index < out_count; ++index)
+        out.push_back(buffer_[index]);
+
+    static auto generator = std::default_random_engine{
+        static_cast<uint32_t>(pseudo_random())};
+    std::shuffle(out.begin(), out.end(), generator);
```

Comment:
Details of \<random\> are purposely buried in random.cpp. The use of a random engine here is redundant with the MT initialization in that location. The MT is statically initialized in order to prevent expensive regeneration. IOW std::shuffle could be initialized as: `std::shuffle(out.begin(), out.end(), get_twister());` and initialization of a new engine should be avoided.

Reply:
I've run into this a number of times actually.  get_twister() isn't exported (statically declared), so it can't be used unless we drop the static function definition and declare it the header.  If you're suggesting that needs to happen, I'm all for that, but every other time I've had to work around it since we generally don't support random anything internally.  But this is a relatively safe case for it ... it seems.

Reply:
We don't implement embedding randomness for the generation of keys, since it's impossible to test and it's very insecure and inflexible.

For randomizing server behavior we rely on random.cpp. We avoid reproducing functionality just because existing is not well factored for a new use. Especially in the case of a new major release we refactor to get the proper generalization.

I looked at the necessary refactor to expose the twister, but we don't actually want to do that, for at least two reasons. One, it creates an abstraction leak. My intent is to not expose randomness types to the code base, so that we have it contained in the random.cpp utils. These functions are too difficult to get right, and they break testability.

So the proper approach is to create a utility that randomizes a collection. The refactor difficulty here is that it requires a template implementation, yet the MT needs to remain private. This is the problem with using functions instead of classes. So it's time to refactor random.cpp into a class.

Reply:
I'm proposing that we use the generator in this case with the following by it:

```
    // FIXME: see discussion on replacing this                                                                                                                          
    // https://github.com/libbitcoin/libbitcoin-network/pull/193                                                                                                        
    static auto generator = std::default_random_engine{
        static_cast<uint32_t>(pseudo_random())};
    std::shuffle(out.begin(), out.end(), generator);
```

And then the libbitcoin re-factor can happen independently and this can be updated then?

Reply:
Code debt, why put it off?

Reply:
Just saw another example of why random.cpp needs to be templatized:

https://libbitcoin.slack.com/archives/C6LH74ACQ/p1520109582000017

Reply:
Is this sort of like what you have in mind?

https://github.com/libbitcoin/libbitcoin/pull/913

Reply:
Yes!